### PR TITLE
Make EKS fetcher to lazy initialize AWS EKS client

### LINKED ISSUE #34949: EKS Auto discovery missing IAM permission stops process execution
## What: 

When the EKS auto-discovery configuration is applied in Teleport and the discovery agent lacks the necessary IAM permission, it prevents the entire Teleport process from starting.

#### Current behavior:

```yaml
discovery_service:
  enabled: yes
  aws:
   - types: ["eks"]
     assume_role_arn: "arn:aws:iam::222222222222:role/example-role"
     regions:
     - eu-central-1
     tags:
       "*": "*"
```

```log
2023-11-24T09:56:46+01:00 DEBU [PROC:1]    Teleport not yet ready: still waiting for DiscoveryReady pid:43228.1 service/supervisor.go:417

ERROR REPORT:
Original Error: *trace.AccessDeniedError AccessDenied: User: arn:aws:iam::11111111111111:user/marek is not authorized to perform: sts:AssumeRole on resource: arn:aws:iam::222222222222:role/example-role
        status code: 403, request id: fe66c920-0df8-4985-9a32-81c1cbc70489
Stack Trace:
        /Users/marek/go/src/github.com/gravitational/teleport/lib/cloud/aws/errors.go:48 github.com/gravitational/teleport/lib/cloud/aws.convertRequestFailureErrorFromStatusCode
        /Users/marek/go/src/github.com/gravitational/teleport/lib/cloud/aws/errors.go:42 github.com/gravitational/teleport/lib/cloud/aws.ConvertRequestFailureError
        /Users/marek/go/src/github.com/gravitational/teleport/lib/cloud/clients.go:1102 github.com/gravitational/teleport/lib/cloud.newSessionWithRole
        /Users/marek/go/src/github.com/gravitational/teleport/lib/cloud/clients.go:651 github.com/gravitational/teleport/lib/cloud.(*cloudClients).getAWSSessionForRole.func1
        /Users/marek/go/src/github.com/gravitational/teleport/lib/utils/fncache.go:135 github.com/gravitational/teleport/lib/utils.FnCacheGet[...].func1
        /Users/marek/go/src/github.com/gravitational/teleport/lib/utils/fncache.go:199 github.com/gravitational/teleport/lib/utils.(*FnCache).get.func1
        /Users/marek/go/go1.21.4/src/runtime/asm_amd64.s:1650 runtime.goexit


Exiting.

Debugger finished with the exit code 0

```

#### Expected  behavior:

Ideally, even if the Teleport discovery_service is missing IAM permissions for EKS auto-discovery, the service should start and trigger other rules for auto-discovery. This behavior should be similar to that of the Database Auto Discovery flow, where missing IAM permissions for AWS DB discovery results in only a warning message to the user.
```yaml
discovery_service:
  enabled: yes
  aws:
  - types: ["redshift"]
    assume_role_arn: "arn:aws:iam::222222222222:role/example-role"
    regions:
    - eu-central-1
    tags:
      "*": "*"
```

```
2023-11-24T10:01:07+01:00 DEBU [DISCOVERY] Skipped fetcher for db at AWS. error:[
ERROR REPORT:
Original Error: *trace.AccessDeniedError AccessDenied: User: arn:aws:iam::11111111111:user/marek is not authorized to perform: sts:AssumeRole on resource: arn:aws:iam::222222222222:role/example-role
        status code: 403, request id: 72af5f4e-a9ca-4507-8176-3994ff4e858a

```

### PR DESCRIPTION
**Pull request by AntonAM**
If one of EKS fetchers initialization encountered an error (for example non authorized to assume IAM role), it stopped the whole discovery process from starting, instead of just skipping this fetcher and logging the problem. This PR changes it so we log warning and continue initialization.

Changelog: Prevent EKS fetcher not having correct IAM permissions from stopping whole Discovery service start up

Fixes #34949

### CODE REVIEW

File: lib/srv/discovery/discovery.go [hunk 1]
```diff
@@ -395,28 +395,10 @@ func (s *Server) initAWSWatchers(matchers []types.AWSMatcher) error {
...
-							Client:       client,
-							Region:       region,
-							FilterLabels: matcher.Tags,
-							Log:          s.Log,
-						},
-					)
-					if err != nil {
-						return trace.Wrap(err)
+						s.Log.WithError(err).Warnf("Could not initialize EKS fetcher(Region=%q, Labels=%q, AssumeRole=%q), skipping.", region, matcher.Tags, matcherAssumeRole.RoleARN)
+						continue
```
Comment (smallinsky) [inline lib/srv/discovery/discovery.go:401]: In this way, the issue where the Teleport service is unable to start will be mitigated, though the UX won't be perfect.

The EKS discovery with missing IAM permissions will be skipped, so if the missing IAM permission is updated, the discovery service requires a restart to apply this change.

Can do "lazy evaluation" when the EKS feature is created   so we will allow to crate a EKS fetcher without full IAM permissions but the evolution will be done in the factual API call ?

Comment (AntonAM) [reply lib/srv/discovery/discovery.go:401]: I've added lazy initialization for the EKS client, so now it will be retrying automatically on every `Get`.

**Comment by AntonAM** [comment]
@fspmarshall @rudream friendly ping

**Review by fspmarshall — APPROVED** [review]

**Review by rudream — APPROVED** [review]

File: lib/cloud/mocks/aws.go [hunk 2]
```diff
@@ -38,11 +38,13 @@ import (
 // STSMock mocks AWS STS API.
 type STSMock struct {
 	stsiface.STSAPI
-	ARN                    string
-	URL                    *url.URL
+	ARN              string
+	URL              *url.URL
+	AssumeRoleErrors map[string]error
```
[the remark this answers is not in the document]
Comment (AntonAM) [reply lib/cloud/mocks/aws.go]: Yep, we don't need it anymore after I changed EKS fetcher to lazy initialization. Removed.

**Pull request merged**